File: rgctl-tests/ecommerce-java/src/main/java/com/example/ecommerce/coolstore/service/CoolstoreOrderService.java
package com.example.ecommerce.coolstore.service;

import com.example.ecommerce.coolstore.model.CoolstoreOrder;
import com.example.ecommerce.coolstore.model.CoolstoreOrderItem;
import com.example.ecommerce.coolstore.model.ShoppingCart;
import com.example.ecommerce.coolstore.model.ShoppingCartItem;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

@Service
public class CoolstoreOrderService {

    private final AtomicLong seq = new AtomicLong(1);
    private final Map<Long, CoolstoreOrder> orders = new ConcurrentHashMap<>();

    public CoolstoreOrder process(ShoppingCart cart) {
        CoolstoreOrder order = new CoolstoreOrder();
        order.setOrderId(seq.getAndIncrement());
        order.setCartId(cart.getCartId());
        order.setCartTotal(cart.getCartTotal());
        List<CoolstoreOrderItem> items = new ArrayList<>();
        for (ShoppingCartItem sci : cart.getShoppingCartItemList()) {
            if (sci.getProduct() != null) {
                items.add(new CoolstoreOrderItem(
                        sci.getProduct().getItemId(),
                        sci.getQuantity(),
                        sci.getPrice()));
            }
        }
        order.setItems(items);
        orders.put(order.getOrderId(), order);
        return order;
    }

    public List<CoolstoreOrder> getOrders() {
        return List.copyOf(orders.values());
    }

    public Optional<CoolstoreOrder> getOrderById(long orderId) {
        return Optional.ofNullable(orders.get(orderId));
    }
}

File: rgctl-tests/ecommerce-cpp/src/coolstore/services/coolstore_order_service.cpp
#include "ecommerce/coolstore/services/coolstore_order_service.hpp"

namespace ecommerce::coolstore {

CoolstoreOrder CoolstoreOrderService::process(ShoppingCart& cart) {
    CoolstoreOrder order;
    order.orderId = seq_++;
    order.cartId = cart.cartId;
    order.cartTotal = cart.cartTotal;
    for (const auto& sci : cart.shoppingCartItemList) {
        if (!sci.hasProduct) continue;
        order.items.emplace_back(sci.product.itemId, sci.quantity, sci.price);
    }
    orders_[order.orderId] = order;
    return order;
}

std::vector<CoolstoreOrder> CoolstoreOrderService::getOrders() const {
    std::vector<CoolstoreOrder> out;
    out.reserve(orders_.size());
    for (const auto& [_, o] : orders_) {
        out.push_back(o);
    }
    return out;
}

std::optional<CoolstoreOrder> CoolstoreOrderService::getOrderById(long orderId) const {
    auto it = orders_.find(orderId);
    if (it == orders_.end()) return std::nullopt;
    return it->second;
}

}  // namespace ecommerce::coolstore
